### ISSUE TITLE
behavior of `aws_profile` is not in line with standard aws tooling

### ISSUE THREAD

**Issue by tbrockmeyer-a** [getsops/sops#1696]
# The Problem

1. Expected behavior for `aws_profile` is that it would use the same credential discovery that the AWS CLI does (or any AWS SDK).
2. In places where the `aws_profile` is not available (in CI or a production server, for instance), the config file is not decryptable without removing the `aws_profile` first.

Most of the `aws_profile` related github issues in this repo revolve around expected behavior vs observed behavior - most of which would be solved by following the same behavior that AWS enforces for their own tooling.

# Proposal

Introduce the correct priority ordering for credential discovery. Instead of forcing the `aws_profile` to be the only method of authentication, we should use the value of `aws_profile` to override the `AWS_PROFILE` environment variable. This would allow the normal credential discovery built into the aws sdk to run its proper course.

# Explanation

### AWS Documentation on Credential Discovery

When using the AWS CLI, there is a [priority](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-configure.html) to which credentials are chosen (I've listed them all here with the relevant ones retaining their descriptions):
> 1. [Command line options](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-options.html) – Overrides settings in any other location, such as the --region, --output, and --profile parameters.
> 2. [Environment variables](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-envvars.html) – You can store values in your system's environment variables.
> 3. Assume role - ...
> 4. Assume role with web identity - ...
> 5. AWS IAM Identity Center - ...
> 6. [Credentials file](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html) – The credentials and config file are updated when you run the command aws configure. The credentials file is located at ~/.aws/credentials on Linux or macOS, or at C:\Users\USERNAME\.aws\credentials on Windows.
> 7. Custom process - ...
> 8. [Configuration file](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html) – The credentials and config file are updated when you run the command aws configure. The config file is located at ~/.aws/config on Linux or macOS, or at C:\Users\USERNAME\.aws\config on Windows.
> 9. Container credentials - ...
> 10. Amazon EC2 instance profile credentials - ...
> 11. IAM Roles for Amazon EC2 - ...

From the AWS Documentation on CLI Credential configuration using environment variables:
> [AWS_PROFILE](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html#cli-configure-files-using-profiles)
> Specifies the name of the AWS CLI profile with the credentials and options to use. This can be the name of a profile stored in a credentials or config file, or the value default to use the default profile.
> 
> If defined, this environment variable overrides the behavior of using the profile named [default] in the configuration file. You can override this environment variable by using the --profile command line parameter.

In other words, the `AWS_PROFILE` environment variable only serves to choose a different profile from the Credentials File or Config File than the [default]. If credentials are present at higher priorities (for example, `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` environment variables), then the `AWS_PROFILE`, credentials file, and config file are all rightfully and completely ignored. Environment Variables have priority 2, the Credentials File has priority 6, and the Config File has priority 8. Never should the contents of the credentials/config file override credentials that are specified with environment variables.

### Sops `aws_profile` Functionality

When a file is encrypted with `sops`, you have the option to specify `--aws-profile`, and that profile is written into the config file as a sort of default profile when decrypting. The most obvious use case of the embedded `aws_profile` appears to be that it allows the use of multiple AWS accounts in a repo without requiring changing environment variables just to edit encrypted files that are encrypted with keys from different accounts. This is a very nice boon for setups where, for instance, non-prod and prod are in separate accounts. Very convenient!

In `sops`, when the `aws_profile` field is set up originally, it is used as a command line argument through `--aws-profile`. Command line arguments are priority 1, so it should override all other settings. This is done correctly by the AWS docs.

When decrypting, it is not a command line argument anymore. It's an implicit override for the `AWS_PROFILE` environment variable. Therefore, if `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` environment variables happen to be available, they should take precedence over it. This can (and has been) observed to not be the case, as `sops` forces the `aws_profile` to be used regardless of any other available credentials - even if the profile is not found at all and there are other valid credentials available.

**Comment by erezhazan1** [comment]
does --aws-profile even work? I'm trying to use it with leapp (which is basically like aws sso cli) and only the default profile works.
I'm trying with this - 
sops --aws-profile "profile" -i ./dir/secret-enc.json

I have in the credentials file 
[profile]
aws_access_key_id=xxx
aws_secret_access_key=xxx
aws_session_token=xxx
region=xxx

**Comment by tbrockmeyer-a** [comment]
As far as I know, it works only when initially creating the file. After that, you have to manually update the value of `"aws_profile"` key in the encrypted file.

**Comment by erezhazan1** [comment]
> As far as I know, it works only when initially creating the file. After that, you have to manually update the value of `"aws_profile"` key in the encrypted file.

oh wow I had no idea that this is how it works, I didn't find anything in docs related to this and I've been days with this issue! 
Thanks so much!

**Comment by jeslinmx** [comment]
> As far as I know, it works only when initially creating the file. After that, you have to manually update the value of `"aws_profile"` key in the encrypted file.

That's not my experience thus far, `--aws-profile` works neither when creating nor subsequently editing for me. On the other hand, `aws_profile:` works both times.

**Comment by erezhazan1** [comment]
> > As far as I know, it works only when initially creating the file. After that, you have to manually update the value of `"aws_profile"` key in the encrypted file.
> 
> That's not my experience thus far, `--aws-profile` works neither when creating nor subsequently editing for me. On the other hand, `aws_profile:` works both times.

you mean `aws_profile:` also if not using it in the first place?

**Comment by felixfontein** [comment]
There has already been a discussion on this a few years back: #634.

One big problem is that for some folks the current way seems absolutely necessary, since they have to use different profiles for different keys. (Example: #1468.)

**Comment by tbrockmeyer-a** [comment]
> One big problem is that for some folks the current way seems absolutely necessary, since they have to use different profiles for different keys. (Example: https://github.com/getsops/sops/issues/1468.)

Reading through this one, I'm not seeing how there's a usage conflict between following the AWS credential discovery flow and using multiple different profiles. For each key, when it is attempting to use it, perform credential discovery, then decrypt/encrypt. Do any of you see how that could not be feasible to implement?

> There has already been a discussion on this a few years back: https://github.com/getsops/sops/issues/634.

Which references the following issues:

#471
> But, when team members share same KMS resource and encrypted files, decryption must be failed if they use different AWS profile name.

And another person in the thread:
> The issue is, the CI server can not assume the developer-role, it can only assume the ci-role. The ideal for me would be a switch like --aws-role where we can override what is in the .sops.yaml file.
The proposal I offered handles these cases, specifically: 

In the case where the `aws_profile` is not available in the Credentials file or Config file, or those files do not exist at all, the `aws_profile` should be ignored entirely to allow those files to still be decrypted.

#614
> Its something to do with --aws-profile - looks like the files have to be decrypted with the same aws-profile name they were encrypted with.

Again, ignoring the missing `aws_profile` would fix this issue by allowing them to use their currently-selected credentials. Even allowing multiple different profiles for the `aws_profile` field (where missing ones in the credentials config are ignored) could solve this problem, where each user of a codebase can put their own profile into the array, allowing everyone to use distinct profiles to take advantage of the functionality.

#619 
> I've upgraded to 3.6.1 and it works.
> Thanks!

This one was a different, unrelated issue, and is fixed already.

#474 
> I'm using version 3.4.0, and seeing this problem when aws_profile is specified in the encrypted file. I'm guessing that aws_profile takes precedence in sops or the AWS credentials provider.

When the `aws_profile` is in the file, if the configuration file is not available, the entire process errors out. If we were to update the code to ignore the `aws_profile` if the configuration file is not available, then this issue is also resolved, as the sdk would find their ec2 credentials.

#439 
> `sops -d test-env.secret.sops.yml --aws-profile production`

This user is expecting the `--aws-profile production` to take precedence over other methods of authenticating when decrypting. This is not yet implemented functionality, and while my proposal does not cover this, my explanation of AWS Credential Discovery does: This user is attempting to pass an authentication method via a CLI argument, which has the highest priority. This should override anything that is in the file, or in environment variables.



In conclusion, I believe that the change I am suggesting can fix all of these issues. The MAIN ISSUE I'm seeing from these posts is that if `aws_profile` exists in the sops file, BOTH the credentials file AND the profile MUST exist on the system it's being run on or else it will error out. There are solutions to this problem, and I'd like to help get it fixed.
